src/Qir/Runtime/stdlib/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#![deny(clippy::all, clippy::pedantic)]
+#![allow(clippy::missing_safety_doc)]
+#![allow(clippy::missing_panics_doc)]
+
+//! # Rust Implementation for Quantum Intermediate Representation
+//! This library implements the classical runtime functions described in the [QIR specification](https://github.com/qir-alliance/qir-spec).
+
+// FUTURE: We should add microbenchmarks to verify behavior of these APIs and have a baseline on how changes affect
+// peformance of the APIs.
+
+pub mod arrays;
+pub mod bigints;
+pub mod callables;
+pub mod math;
+pub mod output_recording;
+pub mod range_support;
+pub mod strings;
+pub mod tuples;
+
+use std::{
+    ffi::CString,
+    rc::{Rc, Weak},
+};
+
+/// Utility used for managing refcounted items.
+unsafe fn update_counts<T>(raw_rc: *const T, update: i32, is_alias: bool) {
+    let mut remaining = update;
+    while remaining != 0 {
+        let rc = Rc::from_raw(raw_rc);
+        if remaining > 0 {
+            // Create and leak new instances to increment the count on the contained item.
+            if is_alias {
+                let _ = Weak::into_raw(Rc::downgrade(&rc));
+            } else {
+                let _ = Rc::into_raw(Rc::clone(&rc));
+            }
+
+            remaining -= 1;
+        } else {
+            // Create and drop instances to decrement the count on contained item.
+            if is_alias {
+                let w = Weak::into_raw(Rc::downgrade(&rc));
+
+                // Need to drop two for a net decrement, since above line increments.
+                drop(Weak::from_raw(w));
+                drop(Weak::from_raw(w));
+            } else {
+                drop(Rc::from_raw(raw_rc));
```

Comment:
Why not use `Rc::decrement_strong_count` and `Rc::increment_strong_count`?

Reply:
Oddly enough, the original prototype for this logic predates the stabilization of those functions by one month... Now that they are stable, I can definitely simplify this logic by relying on them!